File: src/main/java/io/github/gaming32/superpack/FileDialogs.java
package io.github.gaming32.superpack;

import io.github.gaming32.superpack.util.GeneralUtilKt;

import javax.swing.*;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.awt.*;
import java.io.File;

public final class FileDialogs {
    private static final JFileChooser FILE_CHOOSER = new JFileChooser();

    private FileDialogs() {
    }

    public static File mrpack(Component parent) {
        FILE_CHOOSER.setFileSelectionMode(JFileChooser.FILES_ONLY);
        FILE_CHOOSER.setMultiSelectionEnabled(false);
        FILE_CHOOSER.setDialogTitle("Select modpack...");
        FILE_CHOOSER.setAcceptAllFileFilterUsed(false);
        FILE_CHOOSER.resetChoosableFileFilters();
        FILE_CHOOSER.addChoosableFileFilter(new FileNameExtensionFilter("Modrinth Modpacks (*.mrpack)", "mrpack"));
        if (FILE_CHOOSER.showOpenDialog(parent) != JFileChooser.APPROVE_OPTION) {
            return null;
        }
        return FILE_CHOOSER.getSelectedFile();
    }

    public static File saveMrpack(Component parent) {
        FILE_CHOOSER.setFileSelectionMode(JFileChooser.FILES_ONLY);
        FILE_CHOOSER.setMultiSelectionEnabled(false);
        FILE_CHOOSER.setDialogTitle(null);
        FILE_CHOOSER.setAcceptAllFileFilterUsed(false);
        FILE_CHOOSER.resetChoosableFileFilters();
        FILE_CHOOSER.addChoosableFileFilter(new FileNameExtensionFilter("Modrinth Modpacks (*.mrpack)", "mrpack"));
        if (FILE_CHOOSER.showSaveDialog(parent) != JFileChooser.APPROVE_OPTION) {
            return null;
        }
        return GeneralUtilKt.getSelectedSaveFile(FILE_CHOOSER);
    }

    public static File outputDir(Component parent) {
        FILE_CHOOSER.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
        FILE_CHOOSER.setMultiSelectionEnabled(false);
        FILE_CHOOSER.setDialogTitle("Select destination directory...");
        FILE_CHOOSER.setAcceptAllFileFilterUsed(false);
        FILE_CHOOSER.resetChoosableFileFilters();
        if (FILE_CHOOSER.showOpenDialog(parent) != JFileChooser.APPROVE_OPTION) {
            return null;
        }
        return FILE_CHOOSER.getSelectedFile();
    }
}

File: src/main/kotlin/io/github/gaming32/superpack/FileDialogs.kt
package io.github.gaming32.superpack

import com.sun.jna.Platform
import io.github.gaming32.superpack.modpack.ModpackType
import io.github.gaming32.superpack.util.getDownloadsFolder
import io.github.gaming32.superpack.util.toFile
import jnafilechooser.api.JnaFileChooser
import java.awt.Window
import java.io.File
import javax.swing.filechooser.FileNameExtensionFilter

object FileDialogs {
    private val FILE_CHOOSER = JnaFileChooser(getDownloadsFolder())
    private val OUTPUT_FILE_CHOOSER = JnaFileChooser(
        when { // Use the directory containing .minecraft so that people can select it
            Platform.isWindows() -> System.getenv("APPDATA").toFile()
            Platform.isMac() -> File(System.getProperty("user.home"), "Library/Application Support")
            else -> System.getProperty("user.home").toFile()
        }.takeIf { it.isDirectory }
    ).apply {
        mode = JnaFileChooser.Mode.Directories
        isMultiSelectionEnabled = false
        title = "Select destination directory..."
    }

    @JvmStatic
    fun modpack(parent: Window?): File? = FILE_CHOOSER.run {
        mode = JnaFileChooser.Mode.Files
        isMultiSelectionEnabled = false
        title = "Select modpack..."
        addFilter(ModpackType.JOINED_FILE_FILTER)
        ModpackType.values().forEach { addFilter(it.fileFilter) }
        if (showOpenDialog(parent)) {
            null
        } else {
            selectedFile
        }
    }

    @JvmStatic
    fun saveModpack(parent: Window?, type: ModpackType) = FILE_CHOOSER.run {
        mode = JnaFileChooser.Mode.Files
        isMultiSelectionEnabled = false
        title = null
        addFilter(type.fileFilter)
        if (showSaveDialog(parent)) {
            null
        } else {
            selectedFile
        }
    }

    @JvmStatic
    fun outputDir(parent: Window?) = OUTPUT_FILE_CHOOSER.run {
        if (showOpenDialog(parent)) {
            null
        } else {
            selectedFile
        }
    }

    private var JnaFileChooser.title: String?
        get() = throw UnsupportedOperationException()
        set(value) = setTitle(value ?: "")

    private fun JnaFileChooser.addFilter(filter: FileNameExtensionFilter) =
        addFilter(filter.description, *filter.extensions)
}
